### PR TITLE
Allow analyzing code from an open file

### PR DESCRIPTION
**Pull request by Natureshadow**
This implements the feature requested in #100.

### CODE REVIEW

**Comment by Natureshadow** [comment]
@roskakori Anything preventing this from being merged?

File: pygount/analysis.py [hunk 1]
```diff
@@ -706,8 +728,13 @@ def encoding_for(source_path: str, encoding: str = "automatic", fallback_encodin
...
     if encoding == "automatic":
-        with open(source_path, "rb") as source_file:
-            heading = source_file.read(128)
+        if file_handle is None:
+            with open(source_path, "rb") as source_file:
+                heading = source_file.read(128)
+        else:
+            file_handle.seekable and file_handle.seek(0)
+            heading = file_handle.read(128)
+            file_handle.seekable and file_handle.seek(0)
```
Comment (roskakori) [inline pygount/analysis.py:737]: What happens if the `file_handle` is not seekable? Wouldn't this exclude the initial 128 bytes from the analysis?

File: pygount/analysis.py [hunk 2]
```diff
@@ -706,8 +728,13 @@ def encoding_for(source_path: str, encoding: str = "automatic", fallback_encodin
     assert encoding is not None
 
     if encoding == "automatic":
-        with open(source_path, "rb") as source_file:
-            heading = source_file.read(128)
+        if file_handle is None:
+            with open(source_path, "rb") as source_file:
+                heading = source_file.read(128)
+        else:
+            file_handle.seekable and file_handle.seek(0)
```
Comment (roskakori) [inline pygount/analysis.py:735]: What for do we need to `seek(0)` here? Shouldn't the file already be at the beginning?

File: tests/test_analysis.py [hunk 3]
```diff
@@ -198,6 +199,22 @@ def test_can_detect_binary_source_code(self):
         assert source_analysis.state == analysis.SourceState.binary
         assert source_analysis.code_count == 0
 
+    def test_can_analyze_stringio(self):
```
Comment (roskakori) [inline tests/test_analysis.py:202]: Thanks for adding tests. ❤️

File: pygount/analysis.py [hunk 4]
```diff
@@ -261,10 +262,11 @@ def from_file(
         fallback_encoding: str = "cp1252",
         generated_regexes=pygount.common.regexes_from(DEFAULT_GENERATED_PATTERNS_TEXT),
         duplicate_pool: Optional[DuplicatePool] = None,
+        file_handle: Optional[IOBase] = None,
```
Comment (roskakori) [inline pygount/analysis.py:265]: Hm, I had some similar requirement in other applications where a file should be processed either from a path or an already open file. What I usually did was to use a single parameter like this:
```
source_path_or_io: Union[str, pathlib.Path, IO]
```
In case of `str` or `pathlib.Path` the function would open and close the file handle by itself. For everything else, it would just assume an open file-like object and just `read()` it. 

However, one issue arising from allowing file like objects in a single parameter is that not all of them have a proper `name` attribute. That prevents pygments from using the name as basis for figuring out the appropriate lexer.

So your approach with 2 parameters makes sense provided the file name can always be obtained and passed along in your own code calling this function.

So in summary, I guess this is fine.

File: pygount/analysis.py [hunk 5]
```diff
@@ -276,37 +278,52 @@ def from_file(
...
+                    result = SourceAnalysis.from_state(source_path, group, SourceState.error, error)
+            elif not isinstance(file_handle, TextIOBase):
+                if encoding in ("automatic", "chardet"):
+                    encoding = encoding_for(source_path, encoding, fallback_encoding, file_handle=file_handle)
+                try:
+                    source_code = file_handle.read().decode(encoding)
+                except (LookupError, OSError, UnicodeError) as error:
+                    _log.warning("cannot read %s using encoding %s: %s", source_path, encoding, error)
+                    result = SourceAnalysis.from_state(source_path, group, SourceState.error, error)
+            else:
```
Comment (roskakori) [inline pygount/analysis.py:325]: It would be easier to read if this branch is moved upwards and checked with
```python
elif isinstance(file_handle, TextIOBase):
    source_code = file_handle.read()
else:
    ... # former branch from "not isinstance(file_handle, TextIOBase):"

**Review by roskakori — CHANGES_REQUESTED** [review]

File: pygount/analysis.py (hunk 4 above)
Comment (roskakori) [reply pygount/analysis.py:265]: Resolved because it is fine.

File: pygount/analysis.py (hunk 5 above)
Comment (roskakori) [reply pygount/analysis.py:325]: Will be addressed with #106

File: pygount/analysis.py (hunk 1 above)
Comment (roskakori) [reply pygount/analysis.py:737]: Will be addressed with #106

File: pygount/analysis.py (hunk 2 above)
Comment (roskakori) [reply pygount/analysis.py:735]: Will be addressed with #106

**Review by roskakori — APPROVED** [review]
Approving this due to lack of feedback to review comments. The open issues are minor and will be addressed with #106.

**Comment by Natureshadow** [comment]
> Approving this due to lack of feedback to review comments. The open issues are minor and will be addressed with #106.

Oh, soory! I completely forgot about this side project…
